src/domain/entities/DataSet.ts

```diff
@@ -32,6 +39,32 @@ export type CoreCompetency = { id: Id; name: string; code: string };
 export type DataSetList = Pick<DataSetAttrs, "id" | "name" | "lastUpdated" | "permissions">;
 
 export class DataSet extends Struct<DataSetAttrs>() {
+    validateSetup(): Either<ValidationError<DataSet>[], DataSet> {
+        const errors: ValidationError<DataSet>[] = [
+            {
+                property: "name" as const,
+                errors: validateRequired(this.name),
+                value: this.name,
+            },
+            {
+                property: "orgUnits" as const,
+                errors: validateOrgUnits(this.orgUnits),
+                value: this.orgUnits,
+            },
+        ].filter(validation => validation.errors.length > 0);
+
+        return errors.length === 0 ? Either.success(this) : Either.error(errors);
+    }
+
+    updateProject(project: Maybe<Project>): DataSet {
+        const name = project ? `${project.name} DataSet` : "";
+        return this._update({ project, name });
+    }
+
+    update(fieldName: keyof DataSet, value: string | number | boolean): DataSet {
```

Comment:
This is not type-safe. The usual: when you need to reference a type across arguments/output type, you create a type argument: `update<K extends keyof DataSet>(fieldName: K, value: DataSet[K]): DataSet`